### PR TITLE
MariaDB HA via Galera and tests

### PR DESCRIPTION
**Pull request by jcpowermac**
- Adds an HA plan which is provided using Galera.
- Adds testing for HA only (communication to the mariadb service using the other plans does not function needs additional investigation)

### CODE REVIEW

**Comment by jcpowermac** [comment]
@jmontleon can I get a review

cc: @tchughesiv

**Comment by jmontleon** [comment]
@jcpowermac @djzager @jwmatthews @dymurray Overall this looks like a good start to me. I'll try to give it a test drive today.

One concern I have right now this repo serves as the source for both the upstream and downstream apb role packages but this diverges pretty significantly from what we are shipping downstream (and maybe even what we are able to?)

For the mediawiki where we have a different downstream/upstream image we deploy we created a patch file.
https://raw.githubusercontent.com/ansibleplaybookbundle/mediawiki-apb/master/downstream.patch

This gets applied whenever we don't build in the upstream copr:
https://github.com/ansibleplaybookbundle/mediawiki-apb/blob/master/mediawiki-apb-role.spec#L22-L24

I don't have a better idea at the moment as to how we handle this, but it also feels like the diff could get out of hand and become a pain to manage rather quickly in a more advanced case where we're ripping out entire plans as well.

Does anyone have a better idea?

File: playbooks/test.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  - role: ansible.kubernetes-modules
+    install_python_requirements: no
+  - role: ansibleplaybookbundle.asb-modules
+
+  tasks:
+    - include_tasks: test_tasks.yml
+      vars:
+        plan: "{{ outer_item }}"
+      with_items:
+        - ha
```
Comment (djzager) [inline playbooks/test.yml:19]: Why only `ha`? I see you've added `dev`, `ha`, `prod` to playbooks/vars/, but it only looks like you are using the one.

File: playbooks/vars/dev.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+_apb_plan_id: "dev"
+service_name: "rhscl-mariadb"
+namespace: "test-{{ _apb_plan_id }}-{{ service_name }}"
```
Comment (djzager) [inline playbooks/vars/dev.yml:4]: Are we not able to push most of these variables down into the role vars/defaults?

File: roles/rhscl-mariadb-apb-openshift/tasks/test.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+- set_fact:
+    svcobj: "{{ svc.stdout | from_json }}"
+
+- set_fact:
+    mysql_host: "{{ svcobj['spec']['clusterIP'] | default(service_name) }}"
+
+- set_fact:
+    podobj: "{{ pods.stdout | from_json }}"
+
+# Create table using service
```
Comment (djzager) [inline roles/rhscl-mariadb-apb-openshift/tasks/test.yml:19]: Look @dymurray it's a test.yml and I like it.

**Review by djzager — COMMENTED** [review]
I agree with @jmontleon that this is a good start. I had a few comments just from looking it over.

I am a fan of the testing included, although slightly hard to follow between `test_tasks.yml` and separately including `(ha|dev|prod).yml` tasks conditionally on the plan & action variables. I believe we can do better than that.

This is the first example of a test playbook that I have seen that actually does testing instead of simple verifications. @dymurray would be proud.

I'm curious about two things:

1. Could/would/should this be a separate APB entirely with development and production plans for `mariadb-ha`? (I'm not suggesting this is the route we should take, just trying to give possibilities based on @jmontleon 's comments)
1. Is it possible for `ha` to be a parameter instead of a separate plan? We currently have dev|prod plans for mariadb (not ha), why not dev|prod plans for mariadb-ha?

File: playbooks/vars/dev.yml (hunk 2 above)
Comment (djzager) [reply playbooks/vars/dev.yml:4]: I don't believe this is as practical as I originally thought when making the comment, however I do believe the simpler our Ansible code, the better.

File: playbooks/test.yml (hunk 1 above)
Comment (jcpowermac) [reply playbooks/test.yml:19]: @djzager I kept having problems with the tests failing for `dev` and `prod`.  Specifically when trying to the mysql commands to create the table via either the service or pod ip address.  Its something I figured after I got this PR merged that I could look into more.

**Comment by jcpowermac** [comment]
@djzager 

1. I am ok with the ha being in another repo but if that is the case should we do it for the other database apbs as well right?  I am asking this because next up for me is postgresql ha.
2. I was testing with a "dev" like instance for most of my testing so an ephemeral version of HA is certainly doable - the statefulset template would just need to be changed to support that.

File: playbooks/test.yml (hunk 1 above)
Comment (djzager) [reply playbooks/test.yml:19]: That is fair to me. Adding a todo for the remainder would be a good idea then.

**Comment by djzager** [comment]
@jcpowermac 

1. If we can avoid creating new projects to support HA in our database APBs, we should absolutely do that. I know I had mentioned it as a possibility, but your comment about moving onto postgresql made it clear how bad of an idea having 10 database APBs 5 for non-ha and 5 for ha :face_with_thermometer: .
1. I believe it would be best if this (and future) APBs could take the HA as a non-updateable parameter instead of as a plan (it feels like plan vs. parameter is a distinction without a difference).
1. One thing that we need for our APBs, and the Ansible code it is made of, is for them to be as simple and consumable as possible. I notice that we rely heavily on the `when` clause to handle everything from including tasks to deploying k8s services. It is outside the scope of this PR to fix those problems. I would ask though that you find a way not to have `when: action != "test"` throughout the main task file.

File: roles/rhscl-mariadb-apb-openshift/tasks/main.yml [hunk 4]
```diff
@@ -32,12 +34,24 @@
       protocol: TCP
       target_port: 3306
   register: mariadb_service
+  when:
+    - _apb_plan_id != "ha"
+    - action != "test"
```
Comment (djzager) [inline roles/rhscl-mariadb-apb-openshift/tasks/main.yml:39]: I would like it if we could move away from our reliance on conditionals.

File: apb.yml [hunk 5]
```diff
@@ -61,6 +65,15 @@ plans:
     parameters: *_params
     updates_to:
       - prod
+  - name: ha
```
Comment (djzager) [inline apb.yml:68]: Would like for this to be a parameter instead of a plan.

**Review by djzager — CHANGES_REQUESTED** [review]
Putting my earlier comments into a review.

**Comment by jcpowermac** [comment]
@djzager I can start making the changes from a plan -> parameter but that doesn't seem to resolve @jmontleon concerns.

**Comment by djzager** [comment]
:sunglasses: @jcpowermac I knew I was forgetting something.

@jmontleon I have a few questions that I'm not sure how to answer.

1. What is the difference between the redhat and centos images? Why do we need the centos images in order to support ha?

```
    - 'registry.access.redhat.com/rhscl/mariadb-100-rhel7'
    - 'registry.access.redhat.com/rhscl/mariadb-101-rhel7'
    - 'registry.access.redhat.com/rhscl/mariadb-102-rhel7'
    - 'registry.centos.org/centos/mariadb-100-centos7'
    - 'registry.centos.org/centos/mariadb-101-centos7'
    - 'registry.centos.org/centos/mariadb-102-centos7'
```

1. Even if the rhel7 images were compatible, I'm not sure what we would do about https://github.com/jcpowermac/mariadb-apb/blob/9cd265a209c763da32f8a82ea9bdc5dfff8ff898/roles/rhscl-mariadb-apb-openshift/templates/bc-ha.yaml.j2#L11-L16 downstream.

If we are not able to support HA downstream, this appears to require a much more significant patch to the downstream environment than what we have in mediawiki. Maybe we should schedule a meeting to discuss this in detail?

**Comment by jcpowermac** [comment]
@djzager / @jmontleon both are software collection images the only difference is RHEL vs CentOS.  The problem with both is they don't provide the galera packages or `which`.  Also the image needs to be modified so I can template the galera configuration.  That is why there is two buildconfigs.  There is an outstanding PR https://github.com/sclorg/mariadb-container/pull/67 that has gone nowhere fast but that only resolves the first buildconfig.

**Comment by jcpowermac** [comment]
@djzager here is an example of using a dictionary to determine state to reduce the amount of conditionals: https://github.com/jcpowermac/ansible-statemachine-test.  I am not sure this really reduces any complexity.  If this isn't the direction you are thinking it might help if you create an example.

The problem is the database APBs are not simple (dev/prod +ha +update +testing) and the ansible to support all those options is complex.

**Comment by djzager** [comment]
@jcpowermac Thank you for the example. I need to spend some more time thinking about this PR ahead of our call tomorrow.

So that you are aware, I have been working on an [APB reference example](https://github.com/ansibleplaybookbundle/hello-world-apb/pull/3) and one notable omission is a test playbook. After playing around with [`.travis.yml`](https://github.com/djzager/Hello-World-APB/blob/f11efa69949867503604c70bbc9e6c09d64d31b6/.travis.yml) and doing some research, we think it would be great to have a single source of truth for testing APBs (much like [this ansible role test shim](https://gist.github.com/geerlingguy/73ef1e5ee45d8694570f334be385e181) and it being used to test [other ansible roles](https://github.com/geerlingguy/ansible-role-apache/blob/master/.travis.yml) ).

With those things in mind and your excellent use of the test playbook, I need to do some work on the the APB reference example.
